File: Extra/Number_Of_Islands/number__of__islands.py
from typing import Optional, List
from collections import deque


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        height = len(grid)
        width = len(grid[0])
        diff = ((1, 0), (-1, 0), (0, 1), (0, -1))

        count = 0

        def bfs(a, b):
            queue = deque()
            queue.append((a, b))

            while queue:
                y, x = queue.popleft()
                grid[y][x] = count

                for dy, dx in diff:
                    ny, nx = dy + y, dx + x
                    if (
                        0 <= ny < height
                        and 0 <= nx < width
                        and grid[ny][nx] == "1"
                        and (ny, nx) not in queue
                    ):
                        queue.append((ny, nx))

        for h in range(height):
            for w in range(width):
                if grid[h][w] == "1":
                    count += 1
                    bfs(h, w)
        return count

        pass


if __name__ == "__main__":
    sol = Solution()
    grid = [
        ["1", "1", "0", "0", "0"],
        ["1", "1", "1", "0", "0"],
        ["0", "0", "1", "0", "0"],
        ["0", "0", "0", "1", "1"],
    ]
    print(sol.numIslands(grid))
    print("Running Solution...")

File: NeetCode/Graph/Number_Of_Islands/Number_Of_Islands.cpp
// Number_Of_Islands.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    int numIslands(vector<vector<char>> &grid)
    {
        int height = grid.size();
        int width = grid[0].size();
        int count = 0;

        for (int y = 0; y < height; y++)
        {
            for (int x = 0; x < width; x++)
            {
                if (grid[y][x] == '1')
                {
                    count++;
                    bfs({y, x}, grid);
                }
            }
        }
        return count;
    }

    void bfs(pair<int, int> xy, vector<vector<char>> &grid)
    {
        int height = grid.size();
        int width = grid[0].size();
        vector<pair<int, int>> directions = {{0, 1}, {0, -1}, {-1, 0}, {1, 0}};

        queue<pair<int, int>> q;
        q.push(xy);
        grid[xy.first][xy.second] = '#';

        while (!q.empty())
        {
            pair<int, int> tmp = q.front();
            q.pop();

            for (pair<int, int> dydx : directions)
            {
                int ny = tmp.first + dydx.first;
                int nx = tmp.second + dydx.second;
                pair<int, int> newTmp = {ny, nx};
                if (0 <= ny && ny < height &&
                    0 <= nx && nx < width &&
                    grid[ny][nx] == '1')
                {
                    grid[ny][nx] = '#';
                    q.push(newTmp);
                }
            }
        }
    }
};
